src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel_UpdateViewModelState.cs

```diff
@@ -0,0 +1,151 @@
+                             DocumentOutlineHelper.SearchDocumentSymbolData(model.DocumentSymbolData, currentQuery, cancellationToken));
+                    }
+
+                    DocumentSymbolViewModelItems = documentSymbolViewModelItems;
+                }
+
+                if (_currentCaretPosition is { } currentPosition)
+                {
+                    var caretPoint = currentPosition.Point.GetPoint(_textBuffer, PositionAffinity.Predecessor);
+                    if (!caretPoint.HasValue)
+                    {
+                        // document has changed since we were queued to the point that we can't find this position.
+                        return;
+                    }
+
+                    var symbolToSelect = DocumentOutlineHelper.GetDocumentNodeToSelect(DocumentSymbolViewModelItems, model.OriginalSnapshot, caretPoint.Value);
+                    if (symbolToSelect is null)
+                    {
+                        return;
+                    }
+
+                    DocumentOutlineHelper.ExpandAncestors(DocumentSymbolViewModelItems, symbolToSelect.SelectionRangeSpan);
+                    symbolToSelect.IsSelected = true;
+                    _currentExpansionOption = null;
+                }
+            }
+
+            // If we aren't filtering to search results do expand/collapse
+            if (expansion is { } expansionOption && string.IsNullOrEmpty(searchText))
+            {
+                if (expansionOption == ExpansionOption.Collapse && DocumentSymbolViewModelItems.Any())
+                {
+                    _currentExpansionOption = expansionOption;
+                }
+
+                DocumentOutlineHelper.SetExpansionOption(DocumentSymbolViewModelItems, expansionOption);
+            }
+            else if (_currentExpansionOption is ExpansionOption.Collapse)
+            {
+                DocumentOutlineHelper.SetExpansionOption(DocumentSymbolViewModelItems, ExpansionOption.Collapse);
+            }
+
+            static void ApplyExpansionStateToNewItems(ImmutableArray<DocumentSymbolDataViewModel> oldItems, ImmutableArray<DocumentSymbolDataViewModel> newItems)
+            {
+                if (DocumentOutlineHelper.AreAllTopLevelItemsCollapsed(oldItems))
+                {
+                    // new nodes are un-collapsed by default
+                    // we want to collapse all new top-level nodes if 
+                    // everything else currently is so things aren't "jumpy"
+                    foreach (var item in newItems)
+                    {
+                        item.IsExpanded = false;
+                    }
+                }
+                else
+                {
+                    DocumentOutlineHelper.SetIsExpandedOnNewItems(newItems, oldItems);
+                }
+            }
+        }
+
+        private record ViewModelStateData(string? SearchText, CaretPosition? CaretPositionOfNodeToSelect, ExpansionOption? ExpansionOption, bool DataUpdated);
```

Comment:
```suggestion
        private record ViewModelStateDataChange(string? SearchText, CaretPosition? CaretPositionOfNodeToSelect, ExpansionOption? ExpansionOption, bool DataUpdated);
```

Since we create each of these when we're changing something, right?